java-client/src/main/java/org/opensearch/client/json/JsonpMapperBase.java

```diff
@@ -36,9 +36,45 @@
 import jakarta.json.stream.JsonGenerator;
 import jakarta.json.stream.JsonParser;
 import java.lang.reflect.Field;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
 import javax.annotation.Nullable;
 
 public abstract class JsonpMapperBase implements JsonpMapper {
+    @Nullable
+    private Map<String, Object> attributes;
+
+    protected JsonpMapperBase() {}
+
+    protected JsonpMapperBase(JsonpMapperBase o) {
+        this.attributes = o.attributes; // We always copy in `setAttribute` so no need to copy here.
+    }
+
+    @Override
+    public <T> T attribute(String name) {
+        // noinspection unchecked
+        return attributes == null ? null : (T) attributes.get(name);
+    }
+
+    /**
+     * Updates attributes to a copy of the current ones with an additional key/value pair.
+     * Mutates the current mapper, intended to be used in implementations of {@link #withAttribute(String, Object)}
+     */
+    protected JsonpMapperBase addAttribute(String name, Object value) {
+        if (attributes == null) {
+            this.attributes = Collections.singletonMap(name, value);
+        } else {
```

Comment:
Just to skip attributes recreation if the one in question is set already?
```
} else if (!this.attributes.containsKey(name) || !Objects.equals(this.attributes.get(name), value) {
```

Reply:
Have added handling of pre-existing values